Yafc.Model.Tests/Data/DataUtils.cs

```diff
@@ -0,0 +1,414 @@
+
+                formattedValue = DataUtils.FormatAmount(value, unit);
+                Assert.True(DataUtils.TryParseAmount(formattedValue, out parsedValue, unit), $"Could not parse {value} after being formatted as {formattedValue}.");
+                // Even within the allowed range, imprecise formatting is extra imprecise when the value is just over a power of 10; 0.0010209 is formatted as "0.001"
+                // This can't exceed a 5% error: 0.00104999 would also round down to 0.001, but .00105 rounds up.
+                Assert.True(Math.Abs(value - parsedValue) <= Math.Abs(value * .05), $"Incorrectly parsed {value}, formatted as {formattedValue}, to {parsedValue}.");
+            }
+        }
+    }
+
+    [Fact]
+    public void TryParseAmount_IsInverseOfFormatValue_WithBeltsAndPipes() {
+        // Hammer the formatter and parser with lots of random but repeatable values, making sure TryParseAmount can correctly read anything FormatAmount generates.
+        // This time, include b and p suffixes. These suffixes noticably reduce precision, so do them separately.
+        Random r = new Random(0);
+        byte[] bytes = new byte[4];
+        for (int i = 0; i < 1000; i++) {
+            for (UnitOfMeasure unit = 0; unit < UnitOfMeasure.Celsius; unit++) {
+                float value;
+                int count = 1;
+                do {
+                    r.NextBytes(bytes);
+                    value = BitConverter.ToSingle(bytes, 0);
+                    count++;
+                    // TryParseAmount refuses values above 1e15, and FormatAmount has large relative errors for tiny values. But 1e-7 is 1 item every 116 days, so we can ignore that.
+                } while (MathF.Abs(value) is < 1e-7f or > 9.9e14f || float.IsNaN(value));
+
+                Project.current.preferences.itemUnit = r.Next(6) switch {
+                    0 or 1 => 0,
+                    int x => (x - 1) * 15
+                };
+                Project.current.preferences.fluidUnit = r.Next(6) switch {
+                    0 or 1 => 0,
+                    int x => (x - 1) * 60
+                };
+
+                string formattedValue = DataUtils.FormatAmount(value, unit, precise: true);
+                Assert.True(DataUtils.TryParseAmount(formattedValue, out float parsedValue, unit), $"Could not parse {value} after being formatted precisely as {formattedValue}.");
+                // Precise formatting loses a lot of precision when formatting 'N belts' or 'N pipes'.
+                Assert.True(Math.Abs(value - parsedValue) <= Math.Abs(value * .001), $"Incorrectly parsed {value}, formatted precisely as {formattedValue}, to {parsedValue}.");
+
+                formattedValue = DataUtils.FormatAmount(value, unit);
+                // Skip testing if the formatted value is less than 0.1μ; we've lost too much precision for this to be meaningful.
```

Comment:
I guess this is similar/related to #80?

Reply:
Possibly. This test is mostly concerned with formatting and parsing μb(elts) or μp(ipes) of production, though.